File: client/src/model/restaurant.ts
import Cuisine, { isCuisine } from "./cuisine";
import Location, { isLocation } from "./location";
import OpeningTimes, { isOpeningTimes } from "./openingTimes";
import PriceCategory, { isPriceCategory } from "./priceCategory";
import RestaurantLayout, { isRestaurantLayout } from "./restaurantLayout";

/**
 * A Restaurant with the affiliated information
 */
class Restaurant {
    id: string;
    name: string;
    website: string;
    rating: number;
    images: Array<string>;
    priceCategory: PriceCategory;
    cuisines: Array<Cuisine>;
    location: Location;
    layout: RestaurantLayout;
    open: OpeningTimes;


    constructor(
        id: string,
        name: string,
        website: string,
        rating: number,
        images: Array<string>,
        priceCategory: PriceCategory,
        cuisines: Array<Cuisine>,
        location: Location,
        layout: RestaurantLayout,
        open: OpeningTimes
    ) {
        this.id = id
        this.name = name
        this.website = website
        this.rating = rating
        this.images = images
        this.priceCategory = priceCategory
        this.cuisines = cuisines
        this.location = location
        this.layout = layout
        this.open = open
    }

}

export default Restaurant;

/**
 * Checks if an object is a {@link Restaurant}
 * @param restaurant the object to check
 * @returns `true` if the given value is a {@link Restaurant}, `false` otherwise
 */
export const isRestaurant = (restaurant: any): restaurant is Restaurant => {
    return (typeof restaurant === "object") && //
        (typeof restaurant.id === "string") && //
        (typeof restaurant.name === "string") && //
        (typeof restaurant.website === "string") && //
        (typeof restaurant.rating === "number") && //
        (Array.isArray(restaurant.images)) && ((restaurant.images as Array<any>).every(i => typeof i === "string")) && //
        (isPriceCategory(restaurant.priceCategory)) && //
        (Array.isArray(restaurant.cuisines)) && ((restaurant.cuisines as Array<any>).every(isCuisine)) && //
        (isLocation(restaurant.location)) && //
        (isRestaurantLayout(restaurant.layout)) && //
        (isOpeningTimes(restaurant.open));
};

/**
 * Checks if an object is an array of {@link Restaurant}s
 * @param restaruants the object to check
 * @returns `true` if the given value is an array of {@link Restaurant}s, `false` otherwise
 */
export const isRestaurantArray = (restaurants: any): restaurants is Array<Restaurant> => {
    return (Array.isArray(restaurants)) && ((restaurants).every(isRestaurant));
}

File: server/src/main/java/de/nullpointerexception/racetaurant/restaurant/Restaurant.java
package de.nullpointerexception.racetaurant.restaurant;

import java.util.Arrays;
import java.util.Objects;

public record Restaurant(String id, String name, String website, double rating, String[] images,
		PriceCategory priceCategory, Cuisine[] cuisines, Location location, RestaurantLayout layout,
		OpeningTimes times) {

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + Arrays.hashCode(cuisines);
		result = prime * result + Arrays.hashCode(images);
		result = prime * result + Objects.hash(id, layout, location, name, priceCategory, rating, times, website);
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Restaurant other = (Restaurant) obj;
		return Arrays.equals(cuisines, other.cuisines) && Objects.equals(id, other.id)
				&& Arrays.equals(images, other.images) && Objects.equals(layout, other.layout)
				&& Objects.equals(location, other.location) && Objects.equals(name, other.name)
				&& priceCategory == other.priceCategory
				&& Double.doubleToLongBits(rating) == Double.doubleToLongBits(other.rating)
				&& Objects.equals(times, other.times) && Objects.equals(website, other.website);
	}

	@Override
	public String toString() {
		return "Restaurant [id=" + id + ", name=" + name + ", website=" + website + ", rating=" + rating + ", images="
				+ Arrays.toString(images) + ", priceCategory=" + priceCategory + ", cuisines="
				+ Arrays.toString(cuisines) + ", location=" + location + ", layout=" + layout + ", times=" + times
				+ "]";
	}

}
